File: DocumentAnalysisSystem/Visitors/WordCountVisitor.cs
using System.Text.RegularExpressions;
using DocumentAnalysisSystem.Models;

namespace DocumentAnalysisSystem.Visitors
{
    public class WordCountVisitor : IDocumentVisitor
    {
        public int WordCount { get; private set; }

        public void Visit(Document document)
        {
            // No additional word count logic required at document level.
        }

        public void Visit(Section section)
        {
            // Section titles are not counted as document words in this implementation.
        }

        public void Visit(Paragraph paragraph)
        {
            if (string.IsNullOrWhiteSpace(paragraph.Text))
                return;

            var matches = Regex.Matches(paragraph.Text.Trim(), "\\b[\\w']+\\b");
            WordCount += matches.Count;
        }
    }
}

File: DocumentAnalysisSystemJava/src/main/java/com/documentanalysis/visitors/WordCountVisitor.java
package com.documentanalysis.visitors;

import com.documentanalysis.models.Document;
import com.documentanalysis.models.Paragraph;
import com.documentanalysis.models.Section;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class WordCountVisitor implements IDocumentVisitor {
    private static final Pattern WORD_PATTERN = Pattern.compile("\\b[\\w']+\\b");
    private int wordCount;

    public int getWordCount() {
        return wordCount;
    }

    @Override
    public void visit(Document document) {
        // Document-level nodes do not add to the word count.
    }

    @Override
    public void visit(Section section) {
        // Section titles are not counted in this implementation.
    }

    @Override
    public void visit(Paragraph paragraph) {
        if (paragraph.getText() == null || paragraph.getText().isBlank()) {
            return;
        }

        Matcher matcher = WORD_PATTERN.matcher(paragraph.getText().trim());
        while (matcher.find()) {
            wordCount++;
        }
    }
}
